Piechart for player info

ukscid opened the pull request:
Adds pie charts in the player info window

<img width="410" height="187" alt="piechart" src="https://github.com/user-attachments/assets/65d3954e-9da4-4ff9-80cb-1a3fe80cf8e9" />

Commit:
Show pie charts in player info for win loss draw statistic

Commit:
ajust circle for full color on 100%

Commit:
add parameter type to piechart

"%": show only percent value;
else: show name of set, value and percent value

Commit:
extract data from text format, add chart for filter games

benini committed:
Error

benini committed:
Remove ttk::style lookup

Commit:
Bugfix do not try to show pies for missing data

call pinfo from player search has no opponent data

Commit:
debug infos removed

benini commented on tcl/tools/pinfo.tcl:
```diff
@@ -2,6 +2,40 @@
+# Draw Pie Chart
+# Input: window; x,y: upper left corner; name of chart; type: "%": show % only, else show name of set, value and %
+# data: list of list with 3 inputs: name of set, count, color
+proc piechart {w x y width height name type data} {
```
Variable shadowing: the variable "name" below overwrites the procedure parameter name.
Change "name" to something like "title"

benini commented on tcl/tools/pinfo.tcl:
```diff
@@ -2,6 +2,40 @@
+   set start 90
+   ttk_create $w text 0 0 -text $name -anchor nw -tag txt
+   foreach item $data {
+       foreach {name n color} $item break
```
lassign $item name n color

benini commented on tcl/tools/pinfo.tcl:
```diff
@@ -2,6 +2,40 @@
+       if { $extent > 359 } { set extent 359 }
+       $w create arc $coords -start $start -extent $extent -fill $color -outline $color
+       set angle [expr {($start-90+$extent/2)/180.*acos(-1)}]
+       set tx [expr $xm-$rad*sin($angle)]
```
Unbraced expr is slower and potentially unsafe. Several lines have this issue.

benini commented on tcl/tools/pinfo.tcl:
The statistics are shown as wins/draws/losses. For consistency, in the chart as well the wins should come first (on the left) and the losses last (on the right).

benini commented on tcl/tools/pinfo.tcl:
```diff
@@ -2,6 +2,40 @@
+   }
+}
+
 namespace eval pinfo {
```
piechart should not be in the global namespace

benini reviewed (changes requested):
I had the review done by claude-opus, which also had a few other minor notes, but it seems to me that once these details are addressed we can merge.

Commit:
code cleanup

Commit:
cleanup old code

ukscid commented:
fixed the first 3 points.
Order of pie chart should not changed: the chart should run from top
to right sided with wins first.
Maybe piechart could be used for statistics, so it stays in global space.
Maybe the proc should moved to a file in utils.

benini commented:
<img width="501" height="469" alt="image" src="https://github.com/user-attachments/assets/3920360e-5b6c-4617-b5cc-567b763740de" />

The Table Sequence (Left to Right):
First: Wins (+543)
Middle: Draws (=496)
Last: Losses (-313)
The Pie Chart Sequence (Left to Right position):
Left side: Red slice (Losses / -23%)
Center/Bottom: Blue slice (Draws / =37%)
Right side: Green slice (Wins / +40%)
The "Criss-Cross" Problem
Because the table lists Wins first (on the left) but the pie chart places the Green "Win" slice on the right, the mapping is counter-intuitive.

In the top image, the purple arrows have to cross each other to show the correct connection. The arrow for Wins starts on the left and moves to the right; the arrow for Losses starts on the right and moves to the left.

ukscid commented:
I didn't make the connection to the above figures.
From that perspective, it would be correct to swap + and -.
I based my interpretation on how I read pie charts: clockwise. This is like Chessbase shows the pie chart [Statistic](https://players.chessbase.com/de/player/keymer_vincent/126662): the legend reads from left to right, the chart reads clockwise (right to left),
But anyway, feel free to swap wins and losses.

benini committed:
claude-opus version, to be tested

benini committed:
fix_the_fix

benini committed:
Layout Issue Analysis - STILL NOT WORKING

Looking at the widget hierarchy more carefully, I found the real problem in `autoscrollText` and how it's called:

```tcl
autoscrollText both $w.frame $w.text Treeview
```

This creates:
- `$w.frame` - the container frame
- `$w.text` - the text widget as a **sibling** of `$w.frame`, not a child!

Then in `autoscrollBars`:
```tcl
grid $w -in $frame -row $frame_row -column 0 -sticky news
```

The `-in $frame` option positions `$w.text` within `$w.frame`'s space, but since `$w.text` is not actually a child of `$w.frame`, it doesn't clip to the frame's boundaries. The text widget can draw over sibling widgets (the button frames).

The text widget must be a proper child of the frame. Change the call from:

```tcl
autoscrollText both $w.frame $w.text Treeview
```

To:

```tcl
autoscrollText both $w.frame $w.frame.text Treeview
```

Then update **all references** from `$w.text` to `$w.frame.text` in the `playerInfo` proc.

Summary

There are two bugs:

    $w.text configure -state disabled is inside the if { $wlrCount > 36 } block - move it outside
    $w.text is a sibling of $w.frame instead of a child - change to $w.frame.text

The second issue is the root cause of text scrolling over the buttons.

benini committed:
Here's the fully corrected playerInfo proc with all fixes

Looking at this more carefully, I see that the pie chart canvases are also created as children of $w but embedded in the text widget. This could cause clipping issues. They should be children of $w.frame so they're properly contained.

benini committed:
Fix the font typo and probably the intent was to use it

benini committed:
Cleanup (there are still clipping problems)

benini committed:
Cleanup: separate the code that creates the window

benini committed:
almost working

benini committed:
Try with grid

ukscid commented:
Thanks for refactoring pinfo.tcl
Works fine.
If values are 50% 0% 50% then the left text of the first pie chart does not show the first letter. The default value of LABEL_OFFSET should be set to 1.
